toolchain/lc-compile/src/emit.cpp

```diff
@@ -1109,10 +1111,41 @@ void EmitVariableDefinition(intptr_t p_index, PositionRef p_position, NameRef p_
 
 static void EmitForeignHandlerDefinition(intptr_t p_index, PositionRef p_position, NameRef p_name, intptr_t p_type_index, intptr_t p_binding, bool p_force_builtin)
 {
+    MCAutoCustomPointer<struct MCScriptForeignHandlerInfo, MCScriptForeignHandlerInfoRelease> t_info;
+    if (!MCScriptForeignHandlerInfoParse(MCSTR((const char *)p_binding), &t_info))
+    {
+        return;
+    }
+    
+    bool t_force_c_builtin = (*t_info)->language == kMCScriptForeignHandlerLanguageC &&
+                            ForceCBindingsAsBuiltins();
+    
     MCAutoStringRef t_binding_str;
-    if (strcmp((const char *)p_binding, "<builtin>") == 0 || p_force_builtin)
+    if ((*t_info)->language == kMCScriptForeignHandlerLanguageBuiltinC ||
+        p_force_builtin ||
+        t_force_c_builtin)
     {
-        t_binding_str = EmittedBuiltinAdd(p_force_builtin ? nameref_from_cstring((const char *)p_binding) : p_name, p_type_index);
+        NameRef t_symbol_name;
+        if ((*t_info)->language == kMCScriptForeignHandlerLanguageBuiltinC)
+        {
+            t_symbol_name = p_name;
+        }
+        else if (p_force_builtin)
+        {
+            t_symbol_name = nameref_from_cstring((const char *)p_binding);
+        }
+        else
+        {
+            MCAutoStringRefAsCString t_symbol_c_string;
```

Comment:
@runrevmark I guess at this point we should be building a list of unique `(*t_info)->c.library` and writing to stdout at the end for deps